README.md
#  Generic Card Reader Driver Friend

## Introduction

A Lilu plugin that makes System Information recognize your builtin USB card reader as a native one.

![](Assets/Screenshot.png)

## Supported Systems
- macOS Ventura
- macOS Monterey
- macOS Big Sur  
- macOS Catalina  
- macOS Mojave
- macOS High Sierra
- macOS Sierra
- macOS El Capitan

## Dependencies
- [Lilu](https://github.com/acidanthera/Lilu) v1.4.7 or later.

## Limitations & Notes
- This plugin contains a dummy driver that attaches to your USB card reader and injects necessary properties.
- This plugin is derived from but **NOT** compatible with [RealtekCardReaderFriend](https://github.com/0xFireWolf/RealtekCardReaderFriend/).
- This plugin is for decoration purposes only, and your card information will **NOT** be present by System Information.
- This plugin does **NOT** work out of box, and you must follow the [guide](FAQ.md) to specify the vendor and the product identifiers of your card reader.
- If your card reader is powered by either [`AppleUSBCardReader.kext`](file:///System/Library/Extensions/AppleStorageDrivers.kext/Contents/PlugIns/AppleUSBCardReader.kext) or [`RealtekCardReader.kext`](https://github.com/0xFireWolf/RealtekCardReader), you should **NOT** use this plugin.

## Support
If you would like to support my work, please consider a donation.  

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/donate/?business=M6AHXMUVSZQTS&no_recurring=0&item_name=Support+Realtek+PCIe+card+deader+driver+for+macOS&currency_code=USD)


## Boot Arguments
- `-gcrfoff` (or `-liluoff`) to disable
- `-gcrfdbg` (or `-liludbgall`) to enable verbose logging (in DEBUG builds)
- `-gcrfbeta` (or `-lilubetaall`) to enable on macOS newer than 12

## Credits
- [vit9696](https://github.com/vit9696) for [Lilu.kext](https://github.com/vit9696/Lilu)
- [Acidanthera](https://github.com/acidanthera) for [MacKernelSDK](https://github.com/acidanthera/MacKernelSDK)
- [FireWolf](https://github.com/0xFireWolf) for developing the card reader driver and its companion Lilu plugin for macOS

## Acknowledgement
- [dreamwhite](https://github.com/dreamwhite) for testing the early beta release.

## License
This project is licensed under BSD-3-Clause.  
Copyright (C) 2021 FireWolf @ FireWolf Pl. All Rights Reserved.


Changelog.md
#  Changelog

#### v1.0.4
- Add support for macOS Ventura.

#### v1.0.3
- Recomplie the kext with Lilu v1.5.8 to fix the kernel panic of the general protection error.

#### v1.0.2
- Recomplie the kext with Lilu v1.5.7 to fix the kernel panic of the general protection error.

#### v1.0.1
- Set the minimum kernel version to macOS El Capitan. 

#### v1.0.0
- Initial release.


GenericCardReaderFriend/GenericUSBCardReaderController.cpp
//
//  GenericUSBCardReaderController.cpp
//  GenericCardReaderFriend
//
//  Created by FireWolf on 7/24/21.
//

#include "GenericUSBCardReaderController.hpp"

//
// MARK: - Meta Class Definitions
//

OSDefineMetaClassAndStructors(GenericUSBCardReaderController, IOService);

//
// MARK: - IOKit Basics
//

///
/// Start the controller
///
/// @param provider An instance of USB host device that represents the card reader
/// @return `true` on success, `false` otherwise.
///
bool GenericUSBCardReaderController::start(IOService* provider)
{
    // Start the super class
    if (!super::start(provider))
    {
        return false;
    }
    
    // Fetch the USB host device
    this->device = OSDynamicCast(IOUSBHostDevice, provider);
    
    if (this->device == nullptr)
    {
        return false;
    }
    
    this->device->retain();
    
    // Publish the USB device information
    static const char* keys[] =
    {
        "idVendor",
        "idProduct",
        "kUSBSerialNumberString",
    };
    
    OSDictionary* deviceInfo = OSDictionary::withCapacity(3);
    
    for (const char* key : keys)
    {
        deviceInfo->setObject(key, this->device->getProperty(key));
    }
    
    this->setProperty("USB Device Info", deviceInfo);
    
    OSSafeReleaseNULL(deviceInfo);
    
    // Publish the revision level
    this->setProperty("Product Revision Level", "1.00");
    
    // Finished the decoration
    this->registerService();
    
    return true;
}

///
/// Stop the controller
///
/// @param provider An instance of USB host device that represents the card reader
///
void GenericUSBCardReaderController::stop(IOService* provider)
{
    OSSafeReleaseNULL(this->device);
    
    super::stop(provider);
}


GenericCardReaderFriend/GenericUSBCardReaderController.hpp
//
//  GenericUSBCardReaderController.hpp
//  GenericCardReaderFriend
//
//  Created by FireWolf on 7/24/21.
//

#ifndef GenericUSBCardReaderController_hpp
#define GenericUSBCardReaderController_hpp

// Get rid of a bunch of documentation warnings in the USB headers
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdocumentation"
#include <IOKit/usb/IOUSBHostDevice.h>
#pragma clang diagnostic pop

///
/// A generic USB-based card reader controller that injects properties for decoration
///
class GenericUSBCardReaderController: public IOService
{
    //
    // MARK: - Constructors & Destructors
    //
    
    OSDeclareDefaultStructors(GenericUSBCardReaderController);
    
    using super = IOService;
    
    //
    // MARK: - IOKit Basics
    //
    
    /// The USB host device (provider)
    IOUSBHostDevice* device;
    
public:
    ///
    /// Start the controller
    ///
    /// @param provider An instance of USB host device that represents the card reader
    /// @return `true` on success, `false` otherwise.
    ///
    bool start(IOService* provider) override;
    
    ///
    /// Stop the controller
    ///
    /// @param provider An instance of USB host device that represents the card reader
    ///
    void stop(IOService* provider) override;
};

#endif /* GenericUSBCardReaderController_hpp */


GenericCardReaderFriend/GenericCardReaderFriend.cpp
//
//  GenericCardReaderFriend.cpp
//  GenericCardReaderFriend
//
//  Created by FireWolf on 7/24/21.
//

#include <Headers/plugin_start.hpp>
#include <Headers/kern_api.hpp>
#include <Headers/kern_user.hpp>

//
// MARK: - Constants & Patches
//

static const char* kCardReaderReporterPath = "/System/Library/SystemProfiler/SPCardReaderReporter.spreporter";

static const size_t kCardReaderReporterPathLength = strlen(kCardReaderReporterPath);

// MARK: USB-based Card Reader

// Function: SPCardReaderReporter::updateDictionary()
// Find: IOServiceMatching("com_apple_driver_AppleUSBCardReaderSBC")
// Repl: IOServiceMatching("GenericUSBCardReaderController")
// Note: Patch the name of the controller class
static const uint8_t kAppleUSBCardReaderSBC[] =
{
    0x63, 0x6F, 0x6D,                   // "com"
    0x5F,                               // "_"
    0x61, 0x70, 0x70, 0x6C, 0x65,       // "apple"
    0x5F,                               // "_"
    0x64, 0x72, 0x69, 0x76, 0x65, 0x72, // "driver"
    0x5F,                               // "_"
    0x41, 0x70, 0x70, 0x6C, 0x65,       // "Apple"
    0x55, 0x53, 0x42,                   // "USB"
    0x43, 0x61, 0x72, 0x64,             // "Card"
    0x52, 0x65, 0x61, 0x64, 0x65, 0x72, // "Reader"
    0x53, 0x42, 0x43,                   // "SBC"
    0x00                                // "\0"
};

static const uint8_t kGenericUSBCardReaderController[] =
{
    0x47, 0x65, 0x6e, 0x65, 0x72, 0x69, 0x63,
    0x55, 0x53, 0x42,
    0x43, 0x61, 0x72, 0x64,
    0x52, 0x65, 0x61, 0x64, 0x65, 0x72,
    0x43, 0x6f, 0x6e, 0x74, 0x72, 0x6f, 0x6c, 0x6c, 0x65, 0x72,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};

// Function: SPCardReaderReporter::updateDictionary()
// Find: IORegistryEntrySearchCFProperty(entry, "IOService", kCFBundleIdentifierKey, kCFAllocatorDefault, kIORegistryIterateRecursively)
//       NSString isEqualToString("com.apple.driver.AppleUSBCardReader")
// Repl: NSString isEqualToString("science.firewolf.gcrf")
// Note: Patch the bundle identifier
static const uint8_t kAppleUCRBundleIdentifier[] =
{
    0x63, 0x6F, 0x6D,                   // "com"
    0x2E,                               // "."
    0x61, 0x70, 0x70, 0x6C, 0x65,       // "apple"
    0x2E,                               // "."
    0x64, 0x72, 0x69, 0x76, 0x65, 0x72, // "driver"
    0x2E,                               // "."
    0x41, 0x70, 0x70, 0x6C, 0x65,       // "Apple"
    0x55, 0x53, 0x42,                   // "USB"
    0x43, 0x61, 0x72, 0x64,             // "Card"
    0x52, 0x65, 0x61, 0x64, 0x65, 0x72, // "Reader"
    0x00                                // "\0"
};

static const uint8_t kDummyUCRBundleIdentifier[] =
{
    0x73, 0x63, 0x69, 0x65, 0x6E, 0x63, 0x65,
    0x2E,
    0x66, 0x69, 0x72, 0x65, 0x77, 0x6F, 0x6C, 0x66,
    0x2E,
    0x67, 0x63, 0x72, 0x66,
    0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00,
};

//
// MARK: - Helper Functions
//

static inline bool matchReporterPath(const char* path)
{
    return strncmp(path, kCardReaderReporterPath, kCardReaderReporterPathLength) == 0;
}

static void patchReporter(const void* data, vm_size_t size)
{
    void* memory = const_cast<void*>(data);
    
    SYSLOG_COND(UNLIKELY(KernelPatcher::findAndReplace(memory, size, kAppleUSBCardReaderSBC, kGenericUSBCardReaderController)), "GCRF", "Patched the USB controller name.");
    
    SYSLOG_COND(UNLIKELY(KernelPatcher::findAndReplace(memory, size, kAppleUCRBundleIdentifier, kDummyUCRBundleIdentifier)), "GCRF", "Patched the USB bundle identifier.");
}

//
// MARK: - Routed Functions
//

// macOS Catalina and earlier
static boolean_t (*orgCSValidateRange)(vnode_t, memory_object_t, memory_object_offset_t, const void*, vm_size_t, unsigned int*) = nullptr;

static boolean_t wrapCSValidateRange(vnode_t vp, memory_object_t pager, memory_object_offset_t offset, const void* data, vm_size_t size, unsigned int* result)
{
    char path[PATH_MAX];
    
    int pathlen = PATH_MAX;
    
    boolean_t retVal = (*orgCSValidateRange)(vp, pager, offset, data, size, result);
    
    if (retVal && vn_getpath(vp, path, &pathlen) == 0 && matchReporterPath(path))
    {
        patchReporter(data, size);
    }
    
    return retVal;
}

// macOS Big Sur and later
static void (*orgCSValidatePage)(vnode_t, memory_object_t, memory_object_offset_t, const void*, int*, int*, int*) = nullptr;

static void wrapCSValidatePage(vnode_t vp, memory_object_t pager, memory_object_offset_t page_offset, const void* data, int* validated_p, int* tainted_p, int* nx_p)
{
    char path[PATH_MAX];
    
    int pathlen = PATH_MAX;
    
    (*orgCSValidatePage)(vp, pager, page_offset, data, validated_p, tainted_p, nx_p);
    
    if (vn_getpath(vp, path, &pathlen) == 0 && matchReporterPath(path))
    {
        patchReporter(data, PAGE_SIZE);
    }
}

//
// MARK: - Boot Args
//

static const char *bootargOff[] =
{
    "-gcrfoff"
};

static const char *bootargDebug[] =
{
    "-gcrfdbg"
};

static const char *bootargBeta[] =
{
    "-gcrfbeta"
};

//
// MARK: - Plugin Start Routine
//

static KernelPatcher::RouteRequest gRequestLegacy =
{
    "_cs_validate_range",
    wrapCSValidateRange,
    orgCSValidateRange
};

static KernelPatcher::RouteRequest gRequestCurrent =
{
    "_cs_validate_page",
    wrapCSValidatePage,
    orgCSValidatePage
};

static void start()
{
    DBGLOG("GCRF", "Realtek card reader friend started.");
    
    auto action = [](void*, KernelPatcher& patcher) -> void
    {
        KernelPatcher::RouteRequest* request = getKernelVersion() >= KernelVersion::BigSur ? &gRequestCurrent : &gRequestLegacy;
        
        if (!patcher.routeMultipleLong(KernelPatcher::KernelID, request, 1))
        {
            SYSLOG("GCRF", "Failed to route the function.");
        }
    };
    
    lilu.onPatcherLoadForce(action);
}

//
// MARK: - Plugin Configuration
//

PluginConfiguration ADDPR(config) =
{
    xStringify(PRODUCT_NAME),
    parseModuleVersion(xStringify(MODULE_VERSION)),
    LiluAPI::AllowNormal,
    bootargOff,
    arrsize(bootargOff),
    bootargDebug,
    arrsize(bootargDebug),
    bootargBeta,
    arrsize(bootargBeta),
    KernelVersion::ElCapitan,
    KernelVersion::Ventura,
    start
};
